fix: relationunitwatcher data race.

Aflynn50 opened the pull request:
The test for the relation unit watcher had a data race. The notStartedSafeGuard channel was set to nil whilst the pointer could have been in the process of being read by another go routine.

The notStartedSafeGuard was intended to be used as a safeguard to avoid leaking the go-routine that consumed the watcher events in the case of the watcher never starting. This was done by intially leaving the channle open, setting it to nil when the watcher started, and then closing it with a defered finisher (if it was still not nil) at the end of the test. When the channel was nil, it would never be selected, but as soon as it was closed, all recivers would get a nil event. This would allow the consumer go-rountine to exit (with an error).

The race here was that the channel was set to nil when the event consumer started, however, the producer could still be reading the pointer in the channel variable.

It appears that the nil channel was primarily being used as a signal to the finisher, and that an open channel with no events being sent down it would also suffice in this case.

To fix this, I have used a boolean signal to indicate weather the watcher was started. This is not touched from within the go-routines, and allows the finisher to decide weather to free the producer without worrying about weather the channel is nil or not.


## Checklist

<!-- If an item is not applicable, use `~strikethrough~`. -->

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

 go test -race ./... -check.v -check.f TestWatchOneRelationUnit

gfouillet reviewed (dismissed):
**issue:** This fix the datarace but not fix the intent of the startedSafeGuard. This change may cause deadlock if  some other stuff fails.

To fix this issue, we shouldn't change the nature of the channel `notStartedSafeGuard` outside of the loop which listen to it.

However, we can:

* put thing in it
* close it

From another goroutine.

Aflynn50 commented on apiserver/facades/agent/uniter/watcherrelationunit_test.go:
```diff
@@ -314,8 +314,9 @@ func (s *watcherrelationunitSuite) expectWatchRelatedUnitWithEvents(c *gc.C, rel
 	wg := sync.WaitGroup{}
 	notStartedSafeGuard := make(chan struct{})       // used as a safeguard to avoid deadlock if the watcher is not started
 	unexpectedFinishSafeGuard := make(chan struct{}) // used as a safeguard to avoid deadlock if the proxy finish early
+	started := false
```
I don't think this would solve the problem `notStartedSafeGuard` fixes. If the consumer routine was never started, but the producer was, even with the `Add` outside the routines, the first is stuck in its select, with no channels open, and the test deadlocks. We don't just need to keep track of weather the consumer has started, we need to kill it if it hasn't.

Though separately, it would be good to move them outside of the test, if we want our modelling to be correct, so will do that.

SimonRichardson replied:
> I agree that wg.Add should be outside of the goroutine, but you can't get ride of notStartedSafeGuard without risking a deadlock.
> 
> If for some reason (change in the code), the watcher is not registered, the second goroutine which consume event or close unexpectedFinishSafeGuard will never start, which is worse.

If someone doesn't register the watcher, the test will fail. At that point you investigate what you broke.

gfouillet reviewed (approved):
Thank you, I've tried to be too clever, and you outsmart me ;)

SimonRichardson reviewed (changes requested):
Also the way the watcher is registered is incorrect. 

Like https://github.com/juju/juju/blob/b02202a9b5979c1ccb29aa8e2f265b0e728d82cc/apiserver/facades/agent/caasagent/caasagent.go#L127-L138

hpidcock commented:
@Aflynn50 this test is flaky:
```
----------------------------------------------------------------------
FAIL: watcherrelationunit_test.go:159: watcherrelationunitSuite.TestWatchOneRelationUnit

{Changed:map[app1/0:{Version:1} app1/1:{Version:1} app2/0:{Version:1} app2/1:{Version:1}] AppChanged:map[] Departed:[]}
{Changed:map[] AppChanged:map[app1:2 app2:2] Departed:[app1/0 app2/0 app1/1 app2/1]}
watcherrelationunit_test.go:250:
watcherrelationunit_test.go:325:
    c.Check(watcherEvents, gc.DeepEquals, transform.Slice(init.expectedEvents, convertRelationUnitsChange))
... obtained []params.RelationUnitsChange = []params.RelationUnitsChange{params.RelationUnitsChange{Changed:map[string]params.UnitSettings{"app1/0":params.UnitSettings{Version:1}, "app1/1":params.UnitSettings{Version:1}, "app2/0":params.UnitSettings{Version:1}, "app2/1":params.UnitSettings{Version:1}}, AppChanged:map[string]int64(nil), Departed:[]string{}}, params.RelationUnitsChange{Changed:map[string]params.UnitSettings{}, AppChanged:map[string]int64{"app1":2, "app2":2}, Departed:[]string{"app1/0", "app2/0", "app1/1", "app2/1"}}}
... expected []params.RelationUnitsChange = []params.RelationUnitsChange{params.RelationUnitsChange{Changed:map[string]params.UnitSettings{"app1/0":params.UnitSettings{Version:2}, "app1/1":params.UnitSettings{Version:2}, "app2/0":params.UnitSettings{Version:2}, "app2/1":params.UnitSettings{Version:2}}, AppChanged:map[string]int64{"app1":1, "app2":1}, Departed:[]string{}}, params.RelationUnitsChange{Changed:map[string]params.UnitSettings{}, AppChanged:map[string]int64{"app1":2, "app2":2}, Departed:[]string{"app1/0", "app2/0", "app1/1", "app2/1"}}}
... Difference:
...     [0].Changed["app1/0"].Version: 1 != 2
...     [0].Changed["app2/0"].Version: 1 != 2
...     [0].Changed["app1/1"].Version: 1 != 2
...     [0].Changed["app2/1"].Version: 1 != 2
...     [0].AppChanged["app2"]: (missing) != '\x01'
...     [0].AppChanged["app1"]: (missing) != '\x01'


    c.Assert(result, gc.DeepEquals, params.RelationUnitsWatchResult{
        RelationUnitsWatcherId: watcherID,
        Changes:                convertRelationUnitsChange(initialChange),
    })
... obtained params.RelationUnitsWatchResult = params.RelationUnitsWatchResult{RelationUnitsWatcherId:"watcher-id", Changes:params.RelationUnitsChange{Changed:map[string]params.UnitSettings{"app1/0":params.UnitSettings{Version:2}, "app1/1":params.UnitSettings{Version:2}, "app2/0":params.UnitSettings{Version:2}, "app2/1":params.UnitSettings{Version:2}}, AppChanged:map[string]int64{"app1":1, "app2":1}, Departed:[]string{}}, Error:<nil>}
... expected params.RelationUnitsWatchResult = params.RelationUnitsWatchResult{RelationUnitsWatcherId:"watcher-id", Changes:params.RelationUnitsChange{Changed:map[string]params.UnitSettings{"app1/0":params.UnitSettings{Version:1}, "app1/1":params.UnitSettings{Version:1}, "app2/0":params.UnitSettings{Version:1}, "app2/1":params.UnitSettings{Version:1}}, AppChanged:map[string]int64(nil), Departed:[]string{}}, Error:<nil>}
... Difference:
...     Changes.Changed["app1/1"].Version: 2 != 1
...     Changes.Changed["app2/1"].Version: 2 != 1
...     Changes.Changed["app1/0"].Version: 2 != 1
...     Changes.Changed["app2/0"].Version: 2 != 1
...     Changes.AppChanged["app1"]: '\x01' != (missing)
...     Changes.AppChanged["app2"]: '\x01' != (missing)


OOPS: 97 passed, 72 skipped, 1 FAILED
--- FAIL: TestPackage (3.30s)
FAIL
```

gfouillet reviewed (changes requested):
Fix flakiness

gfouillet commented:
@Aflynn50 I double checked the code with the new version. The issue here is that the logic of `expectWatchRelatedUnitWithEvents` should change, since we don't have any guarantee in the order of call for `WatchRelatedUnits` (called in the catacomb loop) and `Register` (called in the main goroutine).

So there is a lot of stuff that can go wrong. 

`expectWatchRelatedUnitWithEvents` should be reworked to reflect this change, but I think it requires mostly simplification (there is less possibility of deadlock with the standard pattern)

Aflynn50 commented:
I've simplified things a good bit, and split the watcher test in two, one to test the watcher with multiple events and one to test just the facade method with only an initial event. I've stress tested this and checked that it doesn't deadlock with artificial failures introduced in the right places.

Aflynn50 commented:
/build

gfouillet commented on apiserver/facades/agent/uniter/uniter.go:
```diff
@@ -1775,20 +1775,23 @@ func (u *UniterAPI) watchOneRelationUnit(
-			RelationUnitsWatcherId: id,
-			Changes:                changes,
-		}, nil
+	id, err := u.watcherRegistry.Register(watch)
```
**issue:** If the firstResult fails, the watcher is still registered. Which is wrong.

Use : https://github.com/juju/juju/blob/48a4d79ccb54ffba3ba829782c01075e1199c8cb/apiserver/internal/watcher.go#L46-L58

(hope your update in test are robust :D)

Aflynn50 replied:
Changed

I've done race and stress tests on the unit tests, as well as manually putting in errors and I can't find anything wrong, so hopefully its ok now.

gfouillet commented on apiserver/facades/agent/uniter/watcherrelationunit_test.go:
```diff
@@ -207,165 +278,68 @@ func (s *watcherrelationunitSuite) TestWatchOneRelationUnit(c *gc.C) {
-	)
+	// Assert:
+	// Start the watcher event consumer, simulating the uniter.
+	wg := sync.WaitGroup{}
```
**todo:** No need of the wg group. This code is equivalent to:

```go

	// Assert: Consume watcher events.
	watcherEvents := make([]params.RelationUnitsChange, 0, len(expectedEvents))
	for v := range relUnitsWatcher.Changes() { // consume all remaining events
		c.Logf("%+v", v)
		watcherEvents = append(watcherEvents, v)
	}
	c.Check(watcherEvents, gc.DeepEquals, transform.Slice(expectedEvents, convertRelationUnitsChange))
```

`unexpectedFinishSafeGuard` can be entirely removed.

Aflynn50 replied:
Good point! That does simplify it a good bit.

I'm not sure we can get rid of the unexpectedFinishSafeGuard though, there is still the possibility the watcher somehow drops some of the changes, and we finish with the go-routine running in the background.

gfouillet replied:
No because you would get stuck in the loop for changes event anyway. The actual code make both safeguard garding to the exact same usecase.

gfouillet reviewed (approved):
LGTM, just apply my todo and verify that everything is right (I have done it locally, but it worth a double check ;)

Aflynn50 commented:
/merge
